### DISCUSSION TITLE
Why get address of variable is unsafe ?

### DISCUSSION THREAD

**Discussion by MBurtsev** [dotnet/csharplang#6182] [Q&A] [answered]
When I use operator for get address(&) compiler shows necessary using unsafe context. But if we just get address and never use it as pointer we are always in safe context. I suggest making changes. So that this requirement applies only if pointers are used for read\write operations. Below I give the code that should work in a safe context.

```C#
        static void Test()
        {
            var adr = 0;
            var threadId = (long)&adr >> 16;
        }
```
Edit:
Another solution is to add an operator like __makeref. For example __stack which will return the top of the call stack(rsp register).

**Comment by MBurtsev** [comment]
I made two benchmarks:

With ManagedThreadId. [class](https://github.com/MBurtsev/DataflowChannel/blob/master/DataflowChannel/ChannelMPOCnoOrder.cs#L110), [bench](https://github.com/MBurtsev/DataflowChannel/blob/master/Benchmark/MPOCnoOrder/MPOCnoOrderWrite.cs)
| Method | Threads |     Mean |    Error |   StdDev |      Min |      Max |         Op/s |     Op/s total |
|------- |-------- |---------:|---------:|---------:|---------:|---------:|-------------:|--------------- |
|  Write |       1 | 11.25 ns | 2.003 ns | 0.520 ns | 10.45 ns | 11.78 ns | 88,884,451.8 |  88,884,451.80 |
|  Write |       2 | 11.88 ns | 1.340 ns | 0.348 ns | 11.49 ns | 12.21 ns | 84,161,901.6 | 168,323,803.20 |
|  Write |       4 | 12.37 ns | 1.732 ns | 0.268 ns | 11.97 ns | 12.54 ns | 80,847,692.9 | 323,390,771.60 |
|  Write |       8 | 15.55 ns | 3.478 ns | 0.538 ns | 14.77 ns | 15.97 ns | 64,320,056.6 | 514,560,452.80 |

Without [class](https://github.com/MBurtsev/Unsafe-ThreadId-Test/blob/master/DataflowChannel/ChannelMPOCnoOrder.cs#L110), [bench](https://github.com/MBurtsev/Unsafe-ThreadId-Test/blob/master/Benchmark/MPOCnoOrder/MPOCnoOrderWrite.cs)
| Method | Threads |      Mean |     Error |    StdDev |       Min |       Max |          Op/s |     Op/s total |
|------- |-------- |----------:|----------:|----------:|----------:|----------:|--------------:|--------------- |
|  Write |       1 |  6.903 ns | 0.3176 ns | 0.0491 ns |  6.845 ns |  6.964 ns | 144,873,381.9 | 144,873,381.90 |
|  Write |       2 |  7.493 ns | 0.2879 ns | 0.0748 ns |  7.392 ns |  7.586 ns | 133,464,121.1 | 266,928,242.20 |
|  Write |       4 |  7.950 ns | 0.4700 ns | 0.1221 ns |  7.736 ns |  8.033 ns | 125,788,900.8 | 503,155,603.20 |
|  Write |       8 | 11.269 ns | 3.0206 ns | 0.4674 ns | 10.685 ns | 11.778 ns |  88,738,969.3 | 709,911,754.40 |

**Reply by yaakov-h** [reply]
How about `Environment.CurrentManagedThreadId` instead?

**Reply by MBurtsev** [reply]
```C#
    public partial class Environment
    {
        public static int CurrentManagedThreadId => Thread.CurrentThread.ManagedThreadId;

```

**Reply by yaakov-h** [reply]
```cs
namespace System
{
    public static partial class Environment
    {
        public static extern int CurrentManagedThreadId
        {
            [MethodImpl(MethodImplOptions.InternalCall)]
            get;
        }
```

https://source.dot.net/#System.Private.CoreLib/src/System/Environment.CoreCLR.cs,0b1681fc8a648379

**Reply by MBurtsev** [reply]
I found it [here](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/mono/System.Private.CoreLib/src/System/Environment.Mono.cs)

### With `Environment.CurrentManagedThreadId`
| Method | Threads |     Mean |    Error |   StdDev |      Min |      Max |         Op/s |     Op/s total |
|------- |-------- |---------:|---------:|---------:|---------:|---------:|-------------:|--------------- |
|  Write |       1 | 11.03 ns | 3.010 ns | 0.782 ns | 10.21 ns | 12.23 ns | 90,690,349.1 |  90,690,349.10 |
|  Write |       2 | 11.66 ns | 2.826 ns | 0.734 ns | 10.79 ns | 12.70 ns | 85,730,947.9 | 171,461,895.80 |
|  Write |       4 | 11.94 ns | 1.546 ns | 0.402 ns | 11.46 ns | 12.51 ns | 83,751,769.7 | 335,007,078.80 |
|  Write |       8 | 15.35 ns | 2.392 ns | 0.621 ns | 14.51 ns | 15.96 ns | 65,156,907.3 | 521,255,258.40 |

### With `Thread.CurrentThread.ManagedThreadId`
| Method | Threads |     Mean |    Error |   StdDev |       Min |      Max |         Op/s |     Op/s total |
|------- |-------- |---------:|---------:|---------:|----------:|---------:|-------------:|--------------- |
|  Write |       1 | 10.56 ns | 4.182 ns | 0.647 ns |  9.599 ns | 11.00 ns | 94,690,726.5 |  94,690,726.50 |
|  Write |       2 | 10.91 ns | 1.902 ns | 0.294 ns | 10.541 ns | 11.23 ns | 91,699,381.2 | 183,398,762.40 |
|  Write |       4 | 11.65 ns | 1.922 ns | 0.297 ns | 11.432 ns | 12.09 ns | 85,851,836.3 | 343,407,345.20 |
|  Write |       8 | 15.02 ns | 2.330 ns | 0.605 ns | 14.155 ns | 15.66 ns | 66,561,493.3 | 532,491,946.40 |

### With stack
| Method | Threads |      Mean |     Error |    StdDev |       Min |       Max |          Op/s |     Op/s total |
|------- |-------- |----------:|----------:|----------:|----------:|----------:|--------------:|--------------- |
|  Write |       1 |  6.978 ns | 0.1525 ns | 0.0396 ns |  6.945 ns |  7.038 ns | 143,316,919.9 | 143,316,919.90 |
|  Write |       2 |  7.423 ns | 0.3070 ns | 0.0475 ns |  7.354 ns |  7.460 ns | 134,711,222.6 | 269,422,445.20 |
|  Write |       4 |  8.151 ns | 0.1704 ns | 0.0443 ns |  8.106 ns |  8.212 ns | 122,677,419.0 | 490,709,676.00 |
|  Write |       8 | 11.772 ns | 1.9715 ns | 0.5120 ns | 10.942 ns | 12.316 ns |  84,947,409.1 | 679,579,272.80 | 

[Sources](https://github.com/MBurtsev/Unsafe-ThreadId-Test/blob/master/DataflowChannel/ChannelMPOCnoOrder.cs#L114)

**Comment by bernd5** [comment]
Can you explain what do you want with that pointer?
The pointer value to the local variable might be always the same... (depending on your current stack)

**Reply by MBurtsev** [reply]
> do you want with that pointer?

Nothing with pointers. But getting the address of a variable in memory can be useful.

> might be always the same...

This can be used to get the thread id. This is useful for making highly efficient algorithms with bounded memory ways.

```
        public unsafe void Write(T value)
        {
            unchecked
            {
                var adr = 0;
                var channel = _channel;
                var id      = (int)((long)&adr >> 16);
                var buffer  = channel.WriterLinks[id % channel.Size];

                if (buffer.Id != id)
                {
                    buffer = SetupThread(id);
                }

                var pos = buffer.WriterPosition;

                if (pos == SEGMENT_CAPACITY)
                {
                    SetNextSegment(channel, id, value);

                    return;
                }

                buffer.WriterMessages[pos] = value;
                buffer.WriterPosition = pos + 1;
            }
        }

```

**Comment by svick** [comment]
I don't think this is sufficient justification to change the language, which is something that requires fairly significant effort, even for a small change like this one.

This change would bring a tiny benefit (not requiring to use `unsafe`) to a tiny number of people (since your use case seems to be extremely narrow). Generally, language changes need to offer significant benefit or affect significant number of people.

And your use case still feels fairly dangerous to me: what guarantees that stacks of two threads can't occupy the same 64 kB block of memory? So I'm not sure not requiring `unsafe` would even be a positive change.

**Reply by MBurtsev** [reply]
> stacks of two threads can't occupy the same 64 kB block

If I create threads, then I know what their stack size is. 

> since your use case seems to be extremely narrow

The areas of multi-threaded programming cannot be called narrow. And it will only grow in the future. There will also be a growing expectation that as cores increase, performance should increase, which is not possible with the standard data structures in net. But the audience of the .Net platform will become narrower due to its inefficiency and limitations.
ConcurrentQueue for example which performance leaves much to be desired. It is used in the basis of asynchronous operations, which is why we saw real freezes in services. Therefore, in our company, part of the services was rewritten in C ++. As time goes by, as C++ becomes even more comfortable, why choose C#? With this mindset, you are burying the future of C#.

**Reply by svick** [reply]
> If I create threads, then I know what their stack size is.

You can decide the stack size, but you can't decide the address where the stack starts, so I still don't see how it's safe.

> The areas of multi-threaded programming cannot be called narrow.

But you're not talking about something that will help many people who are writing multi-threaded code, since it's an incredibly specific pattern. And again, even when it helps, it's a tiny improvement.

I'm all for improving multi-threaded programming in C# (including its performance), but I don't think this suggestion helps with that.

**Reply by En3Tho** [reply]
How do concurrent collections or c++ relate to the topic? (Field access requiring unsafe).  C#/dotnet introduces new perf features/apis with every new release anyway.

As for the topic itself: you're getting the pointer, not a reference to the field. Pointers are unsafe in C#, this isn't something new, really.

**Reply by MBurtsev** [reply]
@svick 
> but you can't decide the address where the stack starts

Two active threads cannot have the same call stack address.

> And again, even when it helps, it's a tiny improvement.

I don't know of any modern project that doesn't use multi-threaded data structures. Right in front of me are 140 repositories with microservices. And that's just my department. Each microservice uses async\await which is built on top of a concurrent queue :)

@En3Tho 
> getting the pointer, not a reference to the field

Only classes(and structures) have fields. Local variables are allocated on the stack. Also, I don't get a pointer, I get an address. Getting the address is safe even for fields since I don't use it for reading or writing.

**Reply by svick** [reply]
> Two active threads cannot have the same call stack address.

Let's say you have two threads with 64 kB stacks. Thread 1 has stack from 0x1234A000 to 0x12359FFF and thread 2 has stack from 0x1235A000 to 0x12369FFF. At a high address on the stack of thread 1, your `threadId` would be 0x1235. At a low address on the stack of thread 2, your `threadId` would also be 0x1235.

> I don't know of any modern project that doesn't use multi-threaded data structures.

But we're not talking about something that will significantly improve those projects. We're talking about avoiding writing `unsafe` couple of times in the implementations of these data structures (assuming they all use your approach to thread IDs).

At a complete guess, this change would save maybe a thousand lines of code *in the whole world*. That doesn't seem worthwhile to me.

**Reply by theunrepentantgeek** [reply]
> Local variables are allocated on the stack.

Not always. They end up on the heap whenever there are captures or async involved (possibly under other circumstances too).

**Reply by MBurtsev** [reply]
> Thread 1 has stack from 0x1234A000 to 0x12359FFF and thread 2 has stack from 0x1235A000 to 0x12369FFF

I already told you that if I create threads with a 64k stack, I won't erase the last two bytes. 

> Not always. They end up on the heap whenever there are captures or async involved (possibly under other circumstances too).

How can a local Int32 variable get into the heap? The heap can only contain objects that have references.

**Reply by CyrusNajmabadi** [reply]
> How can a local Int32 variable get into the heap? 

Nothing in the language requires a local be on the stack (or even that there is a stack at all).

Exposing this is fundamentally unsafe because it's exposing a low level impl detail that the language is under no responsibility to ensure has safe meaning or that it will continue to work in the same way in the future.

**Reply by theunrepentantgeek** [reply]
> How can a local Int32 variable get into the heap? The heap can only contain objects that have references.

I don't want to divert this thread with oodles of example code, so here are two pointers for further research.
* If you have a method decorated with **async** then the body of the method is rewritten by the compiler into a state machine that coordinates the asynchrony, an instance of which lives on the heap. All local variables used by the asynchronous code are lifted onto the class implementing that state machine, and therefore live in the heap.
* If your method contains a lambda expression, any local variables consumed by the body of that lambda expression will similarly be lifted onto a heap dwelling instance; this is required to allow the lambda expression to have a longer lifetime than the original method.

There are also novel implementations of .NET that run without a stack at all - one of those is [HastLayer](https://hastlayer.com/project), for example.

**Reply by MBurtsev** [reply]
@theunrepentantgeek 

I cant use await in unsafe context

**Reply by MBurtsev** [reply]
In any case, if you're not doing read/write operations then getting the address is safe even if it's on the heap.

**Reply by CyrusNajmabadi** [reply]
It's not safe as a pointer 'address' of is not a safe C# concept.

**Reply by CyrusNajmabadi** [reply]
> I cant use await in unsafe context

Right.  And we would not make this start working here.

**Comment by CyrusNajmabadi** [comment]
What's the value in getting the address of something that you don't then intend to read/write. I don't understand the use case.

**Reply by MBurtsev** [reply]
[See usage example.](https://github.com/MBurtsev/Unsafe-ThreadId-Test/blob/master/DataflowChannel/ChannelMPOCnoOrder.cs#L110)

**Reply by CyrusNajmabadi** [reply]
The usage example seems unsafe to me. 

It's also enormously esoteric. I see no reason why requiring `unsafe` here is a problem, esp. given how niche this is.

**Reply by CyrusNajmabadi** [reply]
Afaict, you are reading the address when you are casting it. So this is def something you need to opt into.

**Reply by MBurtsev** [reply]
> The usage example seems unsafe to me.

So what's unsafe? I didn't see any arguments. I'm convinced that it's not even theoretically possible to think of an example where obtaining an address would lead to one of those problems that cause an unsafe context to be used.

**Reply by HaloFour** [reply]
The language considers it safe the moment a pointer is involved.  It doesn't matter that you aren't reading or writing to it, or what it is or isn't pointing to.

**Reply by CyrusNajmabadi** [reply]
What seems unsafe is getting the address of anything and reading it for any purpose. I'm not sure how that could be considered safe from the perspective of the language. The very concept of an address (exposed as a raw pointer) is not safe.

**Reply by MBurtsev** [reply]
The address is just a number. In my example, you do not read what is at this address ). The address becomes a pointer when you use an asterisk (*).  

```C#
var val = 0; 
var ptr = &val; // Here ptr is just number(like long or int) what contains address
*ptr = 10; // Only here we need unsafe context
```

**Reply by CyrusNajmabadi** [reply]
> The address is just a numbe

yes.  A number that is unsafe to get as it's not a defined part of the language.

**Reply by tannergooding** [reply]
> The address becomes a pointer when you use an asterisk (*).

No, all addresses are pointers and raw pointers are unsafe. Safe pointers are GC tracked byrefs such as `ref T`, `in T`, `ref readonly T`, `out T`, other reference types like `class`, `interface` or `delegate`, etc.

Doing almost anything with a raw pointer, regardless of whether you convert it to a number of not is unsafe and often relying on implementation defined behaviors. This is especially true when considering new platforms that may not fit a traditional memory or execution model.

Just use the `unsafe` keyword, there is nothing wrong with requiring it and it won't hurt or hinder your ability to do things.

**Reply by tannergooding** [reply]
Getting the address, even if you simply just treat it like a raw integer may even hurt optimizations because the variable is now considered "address taken" and has an alias that could modify it from almost anywhere.

But additionally, you shouldn't try and use such things for `thread id`. Threads in .NET aren't guaranteed to map to a traditional native threading or stack model. If you want something like a "thread id", you need `Thread.ManagedThreadId` or `Environment.CurrentManagedThreadId`.

The 6ns difference shouldn't matter here as the thread ID should itself be cached elsewhere and not require repeated lookup in the hot loop. Minor tweaks to the API surface cover this as does restructuring your data to use `TLS` (Thread Local Storage) or other utilities designed to improve performance oriented code.

**Reply by MBurtsev** [reply]
> Threads in .NET aren't guaranteed to map to a traditional native threading or stack model

The platform runs on existing hardware. The architecture will not change for the foreseeable future.

> The 6ns difference shouldn't matter

56 million operations per second difference. It may not matter for your purposes. However, do not deny situations when it is important.

>> thread ID should itself be cached elsewhere

How do you cache it if it's the only way to split the threads. I am forced to use this in every operation. The main idea here is that each thread operates in its own area of memory. With this approach, you don't need CAS operations.

**Reply by CyrusNajmabadi** [reply]
> The platform runs on existing hardware. The architecture will not change for the foreseeable future.

That doesn't change whether htis is safe or not.  It's unsafe because it has no guarantees about it preserving meaning as you're stepping outside of the guarantees given.

**Reply by CyrusNajmabadi** [reply]
>  However, do not deny situations when it is important.

Thes question is not if it's important to *you*.  The question is whether it is important enough to warrant a language change.  I'm a language designer of C# and i still haven't gotten an understanding of why using `unsafe` here isa  problem that needs attention.

**Reply by tannergooding** [reply]
> The platform runs on existing hardware.

Yes, including platforms like WASM that do not have a "traditional" execution environment. There is, even today, no guarantee that it maps to the traditional model you're picturing.

> The architecture will not change for the foreseeable future.

The platforms we support and target changes every release and the ecosystem is constantly changing. The runtime makes no guarantees about stack layout or even if it maps to a "traditional" stack setup at all.

> 56 million operations per second difference. It may not matter for your purposes. However, do not deny situations when it is important.

The difference of 6ns is somewhere between `6` (on a 1GHz CPU) and `30` (on a 5GHz CPU) cycles. Even on the most powerful modern hardware, the average memory latency for reads/writes is more than this. This is particularly true in multithreaded environments where you need to force additional `lock#` signals (`Interlocked` operations) and perform additional volatile reads/writes. You're talking about an operation where it will get lost in the noise of the overall system and where a minor re-architecting to avoid repeated lookup of the `id` in your main loop would provide significantly better improvements because you'd be able to remove `12ns` overhead from every call.

> How do you cache it if it's the only way to split the threads. I am forced to use this in every operation. The main idea here is that each thread operates in its own area of memory. With this approach, you don't need CAS operations.

Things like `Thread Local Storage` explicitly use specialized areas of memory and segment registers, namely the `GS` segment register, to help ensure you have cheap thread specific data lookup. Likewise, lookup up the ID once and passing it down into the loop works as well. Because of this, even better than using something like a `thread id` is to just use [ThreadLocal<T>](https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadlocal-1?view=net-6.0) and friends to help reduce the overall cost of maintaining per thread data.

However, if you're not directly using threads and instead are using things like `async`/`await` then using "thread IDs" is incorrect tin the first place. You want [AsyncLocal<T>](https://docs.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?redirectedfrom=MSDN&view=net-6.0) instead which ensures you get task specific locals since a `Task` might migrate between managed threads throughout its lifetime.

**Reply by MBurtsev** [reply]
@tannergooding 
> Even on the most powerful modern hardware

Fast code on weak hardware will be even faster on powerful hardware. Things like cache-friendly code have not been canceled. This comes from my experience. I have a large fleet of test hardware from the Xeon W3690 to the E7 family. Also, you know that asynchronous tasks are just language sugar. Inside are the same threads. Therefore, if you make a queue class, it does not need to know what tasks it is used in, as long as it works correctly with threads.

> The platforms we support and target changes every release and the ecosystem is constantly changing

You are trying to use sophistry. I don't think there will be systems without a stack. Following the ideals and religion, you can believe in anything. But if you look soberly, some things are impossible.

> Thread Local Storage

I decided to check your words. Although I knew the result in advance...
| Method | Threads |     Mean |    Error |   StdDev |      Min |      Max |         Op/s |     Op/s total |
|------- |-------- |---------:|---------:|---------:|---------:|---------:|-------------:|--------------- |
|  Write |       1 | 19.79 ns | 1.154 ns | 0.300 ns | 19.28 ns | 20.02 ns | 50,542,858.1 |  50,542,858.10 |
|  Write |       2 | 20.41 ns | 1.269 ns | 0.196 ns | 20.26 ns | 20.66 ns | 48,989,701.3 |  97,979,402.60 |
|  Write |       4 | 21.46 ns | 2.366 ns | 0.614 ns | 20.85 ns | 22.41 ns | 46,603,714.7 | 186,414,858.80 |
|  Write |       8 | 32.08 ns | 4.393 ns | 1.141 ns | 30.63 ns | 33.06 ns | 31,168,010.0 | 249,344,080.00 |

VS
| Method | Threads |      Mean |     Error |    StdDev |       Min |       Max |          Op/s |     Op/s total |
|------- |-------- |----------:|----------:|----------:|----------:|----------:|--------------:|--------------- |
|  Write |       1 |  6.903 ns | 0.3176 ns | 0.0491 ns |  6.845 ns |  6.964 ns | 144,873,381.9 | 144,873,381.90 |
|  Write |       2 |  7.493 ns | 0.2879 ns | 0.0748 ns |  7.392 ns |  7.586 ns | 133,464,121.1 | 266,928,242.20 |
|  Write |       4 |  7.950 ns | 0.4700 ns | 0.1221 ns |  7.736 ns |  8.033 ns | 125,788,900.8 | 503,155,603.20 |
|  Write |       8 | 11.269 ns | 3.0206 ns | 0.4674 ns | 10.685 ns | 11.778 ns |  88,738,969.3 | 709,911,754.40 |

[Souces](https://github.com/MBurtsev/Unsafe-ThreadId-Test.git)

**Reply by MBurtsev** [reply]
Your advice is 300% slower.

**Reply by HaloFour** [reply]
Fast code that relies on implementation details of a runtime that may not be the same on other runtimes or platforms and is not guaranteed to remain the same even on the current platform is "unsafe", which is what this whole conversation is about.

**Reply by MBurtsev** [reply]
I agree that when the dark elves come and make their PC without a stack, so be it. And also when the processor will not have registers, and when the variables in memory will not have addresses. Only these are such fundamental changes that it seems to me that there will be no .Net platform. :)

**Reply by HaloFour** [reply]
The runtime can emit code as it sees fit.  It doesn't need to use locals or the stack the way you expect it to, even on typical PC hardware.  And the C# language doesn't need to emit locals on the stack either, as is already the case with how it implements coroutines and closures.

But that's fine, there's nothing inherently wrong with "unsafe".  The BCL has its fair share of unsafe code.

**Reply by MBurtsev** [reply]
> doesn't need to emit locals on the stack

`Span<int> numbers = stackalloc[] { 1, 2, 3, 4, 5, 6 };`

Of course not, this safe code will instantiate the Array on the heap.

**Reply by MBurtsev** [reply]
I also used to believe that it was cool to write some abstract code. Why know how hardware works? However, the JIT compiler is far from LLVM. Although it should surpass it. Compiler with artificial intelligence, who knows how to speed up our safe code. It is possible, but who will give money for it? If such reflections, why do we need two classes Queue and ConcurrentQueue. Let the compiler choose the right solution for us. We need to develop not the platform, but the compiler. We will only declare our intentions, the compiler will do the rest. Then the alarm clock rang and I remember that we sometimes get a timeout over SLA of 3000ms, because 32 threads are creating async tasks at the same time, when requests are processed ...

**Reply by CyrusNajmabadi** [reply]
@MBurtsev you still haven't answered my questions on why using `unsafe` is a problem. 

You are not blocked from getting the address. The requirement is simply that you need to opt into this being on by showing that your understand why this is unsafe. Everyone here has given you the reasons why it is not safe (which you seem to accept). So I don't really understand what you're arguing about it asking for. 

You can get her information you need. The language considers this unsafe (which you seem to understand). So you need to use `unsafe`. Why is this a problem?

**Reply by CyrusNajmabadi** [reply]
> We need to develop not the platform, but the compiler. We will only declare our intentions, the compiler will do the rest. 

@MBurtsev this is the repo for the c# *language*, not the compiler. Requests for compiler changes should be made at dotnet/Roslyn (for the c# compiler) or at dotnet/runtime (for the jit,/runtime/aot compilers). 

However, what your are asking for in your OP is a change to the *language*.  Something which you have not only not justified, but where you seem to be ignoring what multiple lang designers are saying. You are then continuing to act as if you are blocked from getting the performance you want. 

You are not blocked. Your are allowed to do this operation. You simply just declare your understanding that this is not safe and you can do it. From my perspective, this is ideal. You get the performance you need, but you understand that it's might break at any point in the future.

**Reply by MBurtsev** [reply]
Friends, I have already come to terms with the fact that I will use unsafe. Thank you all for discussing this issue.

**Reply by CyrusNajmabadi** [reply]
@MBurtsev Your answers are coming across as extremely sarcastic and antagonistic. For example:

> I agree that when the dark elves come and make their PC without a stack, so be it. And also when the processor will not have registers, and when the variables in memory will not have addresses

Please tone it down. This is the repo for the language. It is designed understanding that for now interesting changes have already happened numerous times, so the language should not make assumptions that tie it tightly to one particular impl that behaves one particular way.

**Reply by tannergooding** [reply]
> Your advice is 300% slower.

You're effectively holding it wrong here. You generally wouldn't want to have `class ChannelData` that itself contains multiple `ThreadLocal<T>`. Instead you typically have a single `ThreadLocal<U>` that contains your necessary per-thread data, you'd likewise generally make the `ThreadLocal<U>` data `static` and have it expose the per-thread data at the highest level.

This makes a non-trivial difference in whether you have repeated or one-time lookups. The need to get this "per-thread" data is explicitly meant to be "hoisted" as high as possible to help reduce the overhead and cost. This applies no matter which way you decided to have "per-thread data", even if it was simply some `Dictionary<int, U>`.

Likewise benchmarking multi-threaded code is also very error prone in the first place. Micro-benchmarks run the same code repeatedly in a loop and therefore tend to hit the "best case scenario" for the CPU data, instruction, and microcode caches. This means that it is often "trivial" to write some simple algorithm that appears to perform better than what's built into .NET because the algorithm is effectively only measuring "best case" and is basically tuned explicitly for looking good in a micro-benchmark. The .NET types and code, including `Thread Local Storage` are explicitly written and tuned for real world apps. The best, worst, and average case scenarios are well understood across a range of hardware (from small 1-core, to big many core server systems) and in how the behave in both light, medium, and heavy contention.

**Reply by MBurtsev** [reply]
@tannergooding 
> you'd likewise generally make the ThreadLocal<U> data static

How will a static field work with multiple class instances?

> You're effectively holding it wrong here. You generally wouldn't want to have class ChannelData

Here is TryRead method. For it needs two fields `ChannelData.Head` and `ChannelData.Reader`. And there is no need for others including ThreadLocal. There is also a Clear method, which should clear all data by writing a reference to a new instance with a single atomic operation. Such an organization with the help of ChannelData allows to separate the work of reading and writing threads. To reduce the negative effects of false sharing. Maybe I'm not destined to fully understand your idea without seeing the code...

> Likewise benchmarking multi-threaded code is also very error prone in the first place.
> The best, worst, and average case scenarios are well understood across a range of hardware

I don't believe what you wrote. I believe in the results I get when doing tests. This is the scientific approach. Any statement requires proof. Making statements that cannot be proven are not credible. If you have tests for exotic situations, please provide them, I will also check them. All I see now is some of your beliefs that you pass off as reality. I find it hard to believe without hard evidence that a method that works is 300% faster in the test. In another situation it will be worse. In the same situations, the fast method will work faster than the slow one.

**Comment by theunrepentantgeek** [comment]
> The platform runs on existing hardware. The architecture will not change for the foreseeable future.

So? Language changes need to make sense for every existing platform, not just your narrow special case.

> How do you cache it if it's the only way to split the threads. I am forced to use this in every operation. The main idea here is that each thread operates in its own area of memory. With this approach, you don't need CAS operations.

This sounds like an incredibly narrow special case. What's the value in the language changing for ALL users just to accommodate your esoteric application? 

I'm not denying the value to you in your situation - but languages changes aren't made to suit a single case; they're tremendously expensive to implement and incur an ongoing testing and maintenance burden, so changes need to make sense for a wider audience.

**Reply by HaloFour** [reply]
Not to mention, these assumptions are exactly the reason why the `unsafe` keyword is necessary here.  These are unsafe assumptions.  And the language isn't preventing it, just making you sign off that you understand what you're doing, then it gets out of the way.

**Reply by MBurtsev** [reply]
> esoteric application?

My God, I'm going to make the best data structures in the C# world. Queue, Dictionary and others. Don't tell me that this is esoteric for you and you never use it.

**Reply by CyrusNajmabadi** [reply]
I don't understand why using `unsafe` is a problem to solve those problems.

**Reply by MBurtsev** [reply]
Because it scares a lot of people. And secondly, because simply getting an address is not unsafe :). So far, no one has shown a situation in which this would be dangerous.

**Reply by MBurtsev** [reply]
You can convince me. If you show a couple of lines of code, when you can crash the app without reading and writing. Or make it open to hacking.

**Reply by CyrusNajmabadi** [reply]
> Because it scares a lot of people. 

Why is that relevant?  It's your code.

**Reply by HaloFour** [reply]
> Because it scares a lot of people.

Unless they have to write the code what does it matter?  It's not like you have to mark a method as `unsafe` in order to call an `unsafe` method.

> And secondly, because simply getting an address is not unsafe :). So far, no one has shown a situation in which this would be dangerous.

The language isn't going to split hairs because you think your scenario is "safe enough".  You touch a pointer you're in "unsafe" territory.  There's no reason to expand the heuristics of the compiler to try to figure out if how you're treating that pointer may or may not actually lead to problems.

**Reply by MBurtsev** [reply]
Okay guys I saw your position.

**Comment by AartBluestoke** [comment, marked as answer]
@MBurtsev another way to get the same 'random' distribution is to have something like 
```
public class RoundRobbin{

    private static int _current=0;
    private static int GetNext => System.Threading.Interlocked.Increment(ref _current);
    
    static void Test()
    {
        var threadId = GetNext % Environment.ProcessorCount;
    }
}
```

**Reply by MBurtsev** [reply]
Thanks for your reply. It's a nice idea, but unfortunately it's too slow. I have tested it [before](https://github.com/MBurtsev/DataflowChannel/blob/master/DataflowChannel/Research/ChannelMPMC_B0.cs#L62). The only reason here is that the code is not cache friendly. That's why it's very slow. Also, we cannot use ProcessorCount. This is not thread safe, see the example below.

```C#
    public class RoundRobbin
    {
        const int _count = 100_000_000;
        private int _threads = 4;
        private int _current = 0;
        private int _divider;
        private int[] _storage;

        public RoundRobbin()
        {
            // not thread safe
            //_divider = Environment.ProcessorCount;
            // thread safe. This number must be large enough to use the time lag effect.
            _divider = 1024;
            _storage = new int[_divider];
        }

        public static void Test()
        {
            var ready   = 0;
            var obj = new RoundRobbin();

            for (var n = 0; n < obj._threads; n++)
            {
                Task.Factory.StartNew(() =>
                {
                    for (int i = 0; i < _count; i++)
                    {
                        obj.Add();
                    }

                    if (Interlocked.Add(ref ready, 1) == obj._threads)
                    {
                        obj.Check();
                    }

                }, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
            }
        }

        public void Add()
        {
            var id = Interlocked.Increment(ref _current) % _divider;

            _storage[id]++;
        }

        public void Check()
        {
            var count = 0;

            for (var i = 0; i < _divider; ++i)
            {
                count += _storage[i];
            }

            if (count == _count * _threads)
            {
                Console.Write("It is thread safe");
            }
            else
            {
                Console.Write("It's not thread safe");
            }
        }
    }

```
